### ISSUE TITLE
Eloquent could call `syncChanges()` inside `performInsert()` for accurate "dirty" attributes

### ISSUE THREAD

**Issue by gbradley** [laravel/ideas#1328]
I've been using Eloquent's `isDirty()` to determine which attributes have yet to be saved to the DB. This works well for updates, but I've noticed that, directly after a model is created, the model's attributes are still reported as "dirty".

For example, if I hook into the model lifecycle using the "created" event, I'd expect that no attributes are dirty since the model has just been persisted. However, all attributes are reported as dirty. This seems to be because, while the model's `performUpdate()` method explicitly calls `syncChanges()`, `performInsert()` does not. Both routes end up calling `syncOriginal()`, but this is called after the "created" event is dispatched.

Is there any reason `performInsert()` could not call `syncChanges()`?

**Comment by 36864** [comment]
Can you describe your use-case that involves checking if a newly-created model has dirty attributes? What would that even mean in this context? That somehow a model was changed between being inserted into the database and the event being handled?

**Comment by gbradley** [comment]
Sure. Let's say I have some workflows where I want to check these two specific scenarios:

a) when a model is _created_ and attribute "foo" = 1, then I should change the "bar" attribute to 2.
b) when a model is _updating_ and attribute "foo" _changed_ to 1, then I should send an alert.

To implement this, I do the following:
- observe the "created" event, where I check for the value of the "foo" attribute.
- observe the "updating" event, where I use `isDirty()` to see if "foo" has changed.

What actually happens is:
- the model is created, and if "foo" = 1, then I set "bar" = 2 and call `save()`.
- this save triggers the "updating" event. Because of the behaviour described above, the "foo" attribute is erroneously marked as dirty, and therefore sends an alert.

I can work around this by checking for the `wasRecentlyCreated` property, but it still seems wrong to me that `isDirty()` can return attributes whose values are actually synced.

**Comment by 36864** [comment]
That's fair. I would ask why you're updating after creation instead of intercepting the `creating` event, but I get your point. Should be a relatively simple PR to fix this, I would assume the only reason values aren't synced is because this use case wasn't considered.

**Comment by gbradley** [comment]
Yes, its pretty niche. I'm not using `creating` because in some cases the desired action is to attach models via relationships, but this can't be done before the model is given an ID.
